atomic_reactor/plugins/pre_fetch_sources.py

```diff
@@ -0,0 +1,223 @@
+
+        super(FetchSourcesPlugin, self).__init__(tasker, workflow)
+        self.koji_build = None
+        self.koji_build_id = koji_build_id
+        self.koji_build_nvr = koji_build_nvr
+        self.signing_intent = signing_intent
+        self.koji_fallback = {
+            'hub_url': kojihub,
+            'root_url': kojiroot,
+            'auth': {
+                'proxyuser': koji_proxy_user,
+                'ssl_certs_dir': koji_ssl_certs,
+                'krb_principal': koji_principal,
+                'krb_keytab_path': koji_keytab
+            }
+        }
+        self.session = get_koji_session(self.workflow, self.koji_fallback)
+
+    def run(self):
+        build_id = self.get_image_build_id()
+        signing_intent = self.get_signing_intent(build_id)
+        koji_config = get_koji(self.workflow, {})
+        insecure = koji_config.get('insecure_download', False)
+        urls = self.get_srpm_urls(build_id, signing_intent['keys'], insecure=insecure)
+        sources_list = self.download_sources(urls, insecure=insecure)
+        return sources_list
+
+    def download_sources(self, urls, insecure=False):
+        """Download sources content
+
+        Download content in the guven URLs into a new temporary directory and
+        return a list with each downloaded artifact's path.
+
+        :param urls: int, Koji build id of the container image we want SRPMs for
+        :param insecure: bool, whether to perform TLS checks of urls
+        :return: list, strings with paths to each downloaded file
+        """
+        workdir = tempfile.mkdtemp()
+        dest_dir = os.path.join(workdir, self.DOWNLOAD_DIR)
+        downloaded_sources = []
+
+        for url in urls:
+            parsed_url = urlparse(url)
+            dest_filename = os.path.basename(parsed_url.path)
+            dest_path = os.path.join(dest_dir, dest_filename)
+            self.log.debug('Downloading %s', url)
+
+            if not os.path.exists(dest_dir):
+                os.makedirs(dest_dir)
+
+            req_session = get_retrying_requests_session()
+            request = req_session.get(url, stream=True,
+                                      verify=not insecure)
+            request.raise_for_status()
+
+            with open(dest_path, 'wb') as f:
+                for chunk in request.iter_content(chunk_size=DEFAULT_DOWNLOAD_BLOCK_SIZE):
+                    f.write(chunk)
+
+            self.log.debug('Download finished: %s', dest_path)
+            downloaded_sources.append(dest_path)
+
+        return downloaded_sources
+
+    def get_image_build_id(self):
+        nvr_build_id = None
+        if self.koji_build_nvr:
+            # strict means this raises a koji.GenericError informing no matching build was found in
+            # case the build does not exist
+            self.koji_build = self.session.getBuild(self.koji_build_nvr, strict=True)
+            nvr_build_id = self.koji_build['build_id']
+
```

Comment:
if self.koji_build_id
   check for build and getBuild

and this method can actually return tuple of (build_id, getBuild_result)
that way you don't have to getBuild again in other methods